[IV-24-25] Objetivo 3: Automatización

vpedrosa opened the pull request:
- [x] ¿Incluye el fichero `iv.yaml` las claves de todos los objetivos anteriores?
- [x] ¿Se ha desarrollado el fichero para las tareas a partir de un issue?
- [x] ¿Ese issue está asignado a un milestone?
- [x] ¿El milestone incluye, entre sus productos mínimamente viables, la existencia de herramientas que permitan automatizar tareas?
- [x] ¿Hemos justificado correctamente en caso de elegir un gestor de dependencias el que se use como gestor de tareas?
- [x] ¿El PR, que también es un issue, está asignado al mismo milestone?
- [x] ¿La tarea `check` efectivamente comprueba la sintaxis de los fuentes existentes y sólo eso?
- [x] ¿Se ha comprobado que las instrucciones incluidas en el `README.md` son completas y correctas?

Closes #61
Closes #62
Closes #63
Closes #64
Closes #65

vpedrosa committed:
docs: elijo npm como gestor de dependencias

Documento la decisión de usar npm como gestor de dependencias
para el proyecto TypeScript, comparando con yarn, pnpm y bun
según criterios de compatibilidad, estándar y migración futura.

Closes #61

vpedrosa committed:
docs: elijo npm scripts como gestor de tareas

Documento la decisión de usar npm scripts (package.json) como
gestor de tareas, comparando con Makefile, gulp y grunt según
criterios de idiomaticidad, compatibilidad con bun y simplicidad.

Closes #62

vpedrosa committed:
build: creo package.json con dependencias del proyecto

Declaro typescript como dependencia de desarrollo y añado
el script check para verificar sintaxis del código.

Closes #63

vpedrosa committed:
build: añado clave automatizar a iv.yaml

Indico que el fichero de tareas es package.json y que
las tareas se ejecutan con npm run.

Closes #64

vpedrosa committed:
docs: añado instrucciones de instalación y verificación

Añado secciones de instalación y verificación de sintaxis al README,
y enlace a la documentación de decisiones sobre gestores.

Refs #61, #62, #63, #65

vpedrosa committed:
build: añado package-lock.json

Lockfile generado por npm install para garantizar versiones
reproducibles de las dependencias.

Refs #63

vpedrosa commented:
@GabrielFranciscoSM me echas una mano con una review por favor? 🙏

GabrielFranciscoSM commented:
Ahora le echo un ojo a la documentación, pero antes he mirado los issues. Veo que dentro de estos tratas siempre de definir un problema, que está bien, pero ¿Los títulos reflejan un problema? He visto que haces referencia a los Issues al final de los commits. Está bien, pero ¿no sería más lejible ponerlo antes?

GabrielFranciscoSM commented on docs/gestor.md:
```diff
@@ -0,0 +1,69 @@
+1. **Compatibilidad con el ecosistema**: Debe funcionar con cualquier paquete del registro npm
+2. **Estándar del lenguaje**: Preferencia por herramientas estándar o ampliamente adoptadas
+3. **Migración futura**: Posibilidad de cambiar a otro runtime (bun) sin fricción
+4. **Deuda técnica**: Menor cantidad de configuración adicional requerida
```
¿Es "migración futura" un criterio válido?. Nombras bun, pero no sabes que querrás decidir, y por lo tanto para otro runtime puede no ser compatible.

¿Es la deuda técnica simplemente la cantidad de configuración inicial? Es más, ¿Es la deuda técnica una métrica objetiva?

GabrielFranciscoSM commented on docs/gestor.md:
```diff
@@ -0,0 +1,69 @@
+
+## Decisión
+
+**npm**
```
¿Cómo has tomado esta decisión? ¿Tiene sentido tomar una decisión antes de evaluar las alternativas?

GabrielFranciscoSM commented on docs/gestor.md:
```diff
@@ -0,0 +1,69 @@
+## Justificación
+
+- Es el **estándar de facto** del ecosistema Node.js/TypeScript, incluido en cualquier instalación de Node
+- El fichero `package.json` que genera es **compatible con todos los demás gestores** (yarn, pnpm, bun)
```
¿El fichero lo genera npm? ¿Este fichero también lo pueden generar los otros runtimes? En caso afirmativo, ¿El criterio sirve de algo?

GabrielFranciscoSM commented on docs/gestor.md:
```diff
@@ -0,0 +1,69 @@
+- Es el **estándar de facto** del ecosistema Node.js/TypeScript, incluido en cualquier instalación de Node
+- El fichero `package.json` que genera es **compatible con todos los demás gestores** (yarn, pnpm, bun)
+- **Cero configuración adicional**: no requiere instalar herramientas extra
+- Permite **migrar a bun** en el futuro simplemente cambiando el comando (`npm install` → `bun install`)
```
Si tu criterio es migrar a bun, ¿Por qué no elegir bun?

GabrielFranciscoSM reviewed (commented):
Hola, disculpa por la tardanza. En general mis comentarios van en la linea de que una buena justificación debe de reflejar un correcto uso de la metodología, porque en este documento es únicamente dónde se refleja cómo lo has escrito.

A parte, aunque no lo pone en la página del objetivo 3, en el caso del lenguaje javascript/typescript hay que elegir un runtime y justificar la decisión como con los gestores.

P.D. En cuanto a estilo, usar varios títulos principales (#) no es una buena práctica, ya que sería como usar varios <h1>.

vpedrosa commented:
Buenas @GabrielFranciscoSM, gracias por tu revisión. Estoy haciendo los cambios.
Sobre esto:

> Ahora le echo un ojo a la documentación, pero antes he mirado los issues. Veo que dentro de estos tratas siempre de definir un problema, que está bien, pero ¿Los títulos reflejan un problema? He visto que haces referencia a los Issues al final de los commits. Está bien, pero ¿no sería más legible ponerlo antes?

He modificado los títulos de los issues, efectivamente eran más la solución que el problema. El tema de referenciar las issues en los commits al final es por convención, pero sí puede ser más legible ponerlo antes. Cambio el formato para futuros commits.

vpedrosa committed:
Refs #61, #62 - docs: mejoro justificación de elección de runtime y gestores

Reestructuro docs/gestor.md con criterios objetivos y medibles:
- Añado sección de elección de runtime (Node.js, Bun, Deno)
- Uso tablas de evaluación antes de cada decisión
- Elimino criterios vagos (migración futura, deuda técnica, simplicidad)
- Corrijo estructura markdown (un solo # como título)

vpedrosa commented:
@GabrielFranciscoSM creo que lo tengo mejor, ¿cómo lo ves?

GabrielFranciscoSM reviewed (commented):
Lo veo bastante bien. También me gustaría ver alguna referencia del origen de las métricas, por ejemplo de la compatibilidad de paquetes npm con bun.

vpedrosa committed:
Refs #61 - docs: añado referencias a métricas de runtime

- Añado enlace a documentación de LTS de Node.js
- Añado enlace a compatibilidad npm de Bun
- Elimino mención a rendimiento (no es criterio definido)

vpedrosa commented:
@JJ solicito revisión, por favor.

JJ commented on docs/gestor.md:
```diff
@@ -0,0 +1,87 @@
+|----------|-------------|--------------|
+| Madurez | Tiempo en producción y estabilidad de la API | Años desde versión 1.0 |
+| Soporte LTS | Mantenimiento a largo plazo garantizado | Sí/No y duración |
+| Compatibilidad npm | Capacidad de usar paquetes del registro npm | Porcentaje de compatibilidad |
```
También es arbitrario. Y en todo caso todos lo cumplen. Con lo que no es compatible deno es con el gestor de dependencias, pero [puede usar paquetes npm especificándolos como npm](https://deno.com/npm):

JJ reviewed (commented):
Ha mejorado bastante, pero aún le queda. En clase mencionamos Snyk Advisor que da una puntuación "de salud" unitaria de herramientas en diferentes lenguajes.

vpedrosa committed:
Refs #61, #62 - docs: reescribo justificación eligiendo Bun

- Añado velocidad como criterio de decisión
- Amplío gestores de dependencias a 6 alternativas con Snyk Advisor
- Amplío gestores de tareas a 9 alternativas
- Cambio decisión de Node.js+npm a Bun (runtime+dependencias+tareas)
- Uso formato de requisitos y análisis similar a PRs aprobadas

vpedrosa commented:
He revisado cómo lo ha hecho @GabrielFranciscoSM que ya tiene la PR aprobada y he reescrito la documentación siguiendo un formato similar.

También he incorporado Snyk Advisor como me recomendó @JJ. He ampliado el número de alternativas analizadas y he añadido velocidad como criterio, y al final he acabado eligiendo Bun sobre Node.js, y eso ha hecho que al final cambie la decisión de gestor de dependencias y gestor de tareas.

vpedrosa committed:
Closes #63, #64 - build: migro proyecto a Bun

- Actualizo iv.yaml para usar bun run
- Cambio package.json a type: module
- Actualizo README.md con instrucciones de Bun
- Elimino package-lock.json (Bun usa bun.lockb)

JJ reviewed (commented):
1. Cambiar totalmente el envío, y más a lo que ha elegido un compañero ya, en vez de cambios específicos a la retroalimentación dada no es nunca una buena idea.
2. En el caso de Javascript la elección de un runtime lleva aparejada la elección de un fichero para especificar las dependencias, y eso a su vez excluye otros gestores de dependencias que no lo usen.

vpedrosa committed:
Refs #61, #62 - docs: añado enlaces y elimino Deno de gestores

- Añado enlaces a documentación oficial de todas las herramientas
- Elimino Deno de gestores de dependencias y tareas (requiere otro runtime)

vpedrosa commented:
He modificado el envío porque tras ver lo que había hecho Gabriel (también me fijé en José Antonio), me costaba defender Node y NPM, porque realmente el criterio principal para elegirlos (aunque no lo había puesto), es porque tengo más experiencia con ellos, así que como comentaste, al final había hecho una elección "amañada" para que salieran esos. Por eso he creído necesario cambiar el formato y la entrega completa.

He modificado con los comentarios que me has indicado. @JJ Listo para revisión.

vpedrosa committed:
Refs #61, #62 - docs: simplifico gestor de dependencias

La elección del runtime determina el gestor de dependencias disponible.
Al elegir Bun, el gestor viene integrado.

vpedrosa commented:
Solicito revisión, por favor, @JJ

JJ commented:
Sería mejor que algún compañero revisara de nuevo antes. Me da la impresión de que has corregido sólo puntualmente lo que te he comentado, pero hay otras partes del envío que tienen el mismo problema (es decir: son herramientas que no hacen lo que dices que hacen) pero no las has corregido.

GabrielFranciscoSM commented:
@vpedrosa en cuanto mires lo que te comenta JJ puedes solicitarme revisión para echarle un ojo.

vpedrosa committed:
Refs #61, #62 - docs: corrijo descripción de herramientas en gestor de tareas

- Amplio descripción del gestor de tareas (no solo sintaxis, también tests y build)
- Añado puntuaciones de Snyk Advisor a todas las alternativas
- Corrijo datos de gulp: 84/100 con bajo mantenimiento (6 commits/12 meses)
- Elimino Task (no disponible en npm)
- Elimino Make (no específico del ecosistema)
- Actualizo justificación final con referencias a las puntuaciones

vpedrosa commented:
He modificado un poco el análisis teniendo en cuenta todo lo que había, @GabrielFranciscoSM cuando puedas échale un vistazo. Gracias!

GabrielFranciscoSM commented on docs/gestor.md:
```diff
@@ -0,0 +1,69 @@
+        - Cumple el requisito de compatibilidad npm (100% al ser el runtime original del ecosistema). Es el runtime más maduro (desde 2009) con el mayor número de recursos y documentación disponibles. Ofrece [soporte LTS oficial de 30 meses por versión](https://nodejs.org/en/about/previous-releases). En cuanto a velocidad, no destaca frente a alternativas más nuevas.
+
+    * **[`Bun`](https://bun.sh/)**
+        - Cumple todos los requisitos. Tiene [compatibilidad 100% con npm según su documentación](https://bun.sh/docs/runtime/nodejs-apis). Destaca claramente en velocidad: es significativamente más rápido que Node.js en ejecución, instalación de dependencias y arranque. Según [Snyk Advisor](https://snyk.io/advisor/npm-package/bun), tiene un health score de 87/100 con mantenimiento muy activo. Al ...
```
En la documentación no especifica que sea 100% compatible.

GabrielFranciscoSM commented on docs/gestor.md:
```diff
@@ -0,0 +1,69 @@
+        - Cumple todos los requisitos. Tiene [compatibilidad 100% con npm según su documentación](https://bun.sh/docs/runtime/nodejs-apis). Destaca claramente en velocidad: es significativamente más rápido que Node.js en ejecución, instalación de dependencias y arranque. Según [Snyk Advisor](https://snyk.io/advisor/npm-package/bun), tiene un health score de 87/100 con mantenimiento muy activo. Al ...
+
+    * **[`Deno`](https://deno.com/)**
+        - Cumple parcialmente los requisitos. Tiene [compatibilidad con npm usando el prefijo `npm:`](https://docs.deno.com/runtime/fundamentals/node/) en las importaciones, pero usa un fichero de dependencias diferente (`deno.json`). Ofrece TypeScript nativo sin configuración adicional y un modelo de seguridad por permisos. En velocidad es competitivo pero no alcanza a Bun. Su ecosistema de herr ...
```
¿Es un requisito que ofrezca typescript nativo? ¿Aporta valor a la decisión incluirlo?

¿Cómo sabes que Deno es rápido pero no tanto como Bun?

GabrielFranciscoSM commented on docs/gestor.md:
```diff
@@ -0,0 +1,69 @@
+
+## Elección del gestor de dependencias
+
+Al haber elegido Bun como runtime, el gestor de dependencias viene integrado. Usa `package.json` y genera `bun.lock` como lockfile. Es significativamente más rápido que npm y pnpm en instalación de dependencias.
```
Aunque es cierto que bun incluye gestor de tareas, ¿no deberías de valorar y/o estudiar alternativas?

Es más, ¿crees que es adecuado que las decisiones dependan de elecciones anteriores?

GabrielFranciscoSM commented on docs/gestor.md:
```diff
@@ -0,0 +1,69 @@
+2. **Análisis de alternativas**:
+
+    * **[`Bun`](https://bun.sh/)**
+        - Cumple los requisitos. Según [Snyk Advisor](https://snyk.io/advisor/npm-package/bun), tiene un health score de 87/100 con mantenimiento activo (última release hace 7 días, último commit hace 2 días). Usa directamente el archivo `package.json`, lo que ofrece simplicidad para crear tareas. Facilita la integración con GitHub Actions. Coherente con la elección de Bun como runtime.
```
¿Es un criterio el tipo de archivo de gestión que se use y lo simple que sea?

GabrielFranciscoSM commented on docs/gestor.md:
```diff
@@ -0,0 +1,69 @@
+        - Cumple los requisitos. Según [Snyk Advisor](https://snyk.io/advisor/npm-package/bun), tiene un health score de 87/100 con mantenimiento activo (última release hace 7 días, último commit hace 2 días). Usa directamente el archivo `package.json`, lo que ofrece simplicidad para crear tareas. Facilita la integración con GitHub Actions. Coherente con la elección de Bun como runtime.
+
+    * **[`Nx`](https://nx.dev/)**
+        - Cumple los requisitos. Según [Snyk Advisor](https://snyk.io/advisor/npm-package/nx), tiene un health score de 89/100 con mantenimiento muy activo (última release hace 5 días). Tiene un archivo de configuración `nx.json` para configurar paralelismo, caché y otras opciones. Ofrece buena integración CI con GitHub, GitLab y otras plataformas. Orientado a proyectos grandes, añade complejidad ...
```
¿Igual que antes, son criterios que ofrezca paralelismo, caché y otras opciones?

GabrielFranciscoSM commented on docs/gestor.md:
```diff
@@ -0,0 +1,69 @@
+        - Cumple los requisitos. Según [Snyk Advisor](https://snyk.io/advisor/npm-package/nx), tiene un health score de 89/100 con mantenimiento muy activo (última release hace 5 días). Tiene un archivo de configuración `nx.json` para configurar paralelismo, caché y otras opciones. Ofrece buena integración CI con GitHub, GitLab y otras plataformas. Orientado a proyectos grandes, añade complejidad ...
+
+    * **[`Turborepo`](https://turborepo.com/)**
+        - Cumple los requisitos. Según [Snyk Advisor](https://snyk.io/advisor/npm-package/turbo), tiene un health score de 94/100 con mantenimiento muy activo (última release hace 4 días). Tiene un archivo de configuración `turbo.json` para configurar paralelismo, tareas y dependencias. Buena integración con plataformas CI y facilita el desarrollo en local y remoto.
```
Igual

GabrielFranciscoSM commented on docs/gestor.md:
```diff
@@ -0,0 +1,69 @@
+
+**Decisión: Bun**
+
+Nx (89/100) y Turborepo (94/100) tienen muy buena puntuación en Snyk Advisor, pero requieren configuración adicional (`nx.json`, `turbo.json`) que añade complejidad innecesaria. gulp (84/100) tiene bajo mantenimiento (6 commits en 12 meses) y requiere dependencias adicionales. Se elige Bun (87/100) como gestor de tareas por su buen mantenimiento y coherencia con el runtime elegido, manteniendo un ...
```
¿La configuración adicional es un criterio?

GabrielFranciscoSM reviewed (commented):
Hola. Te he comentado unas cuantas cosas pero la mayoría reflejan el mismo problema.

vpedrosa committed:
Refs #61, #62 - docs: añado análisis completo de gestores de dependencias

Se evalúan npm, Yarn y Bun con requisitos definidos:
- npm (89/100): cumple mantenimiento pero es más lento
- Yarn (56/100): no cumple por mantenimiento inactivo
- Bun (92/100): cumple todos los requisitos

vpedrosa commented:
Muchas gracias @GabrielFranciscoSM por la revisión, creo que ya está mejor. @JJ si puedes revisa los cambios que he realizado. Muchas gracias y feliz navidad a ambos!

JJ commented on docs/gestor.md:
```diff
@@ -0,0 +1,84 @@
+    * **[`Nx`](https://nx.dev/)**
+        - Cumple los requisitos. Según [Snyk Advisor](https://snyk.io/advisor/npm-package/nx), tiene un health score de 89/100 con mantenimiento muy activo (última release hace 5 días). Tiene un archivo de configuración `nx.json` para configurar opciones. Ofrece buena integración CI con GitHub, GitLab y otras plataformas.
+
+    * **[`Turborepo`](https://turborepo.com/)**
```
En [este comentario](https://github.com/vpedrosa/wave-planner/pull/66#issuecomment-3685548926) te dije que revisaras otras herramientas para que efectivamente hicieran lo que tú dices que hacen. ¿Has revisado esta?

vpedrosa replied:
Vale, perdona que me había pensado que te referías a otra cosa. Es cierto.

JJ commented on package.json:
```diff
@@ -0,0 +1,12 @@
+  "description": "Detector de buenas épocas de surf para playas",
+  "type": "module",
+  "scripts": {
+    "check": "bun run tsc --noEmit"
```
¿Seguro que necesitas ejecutar tsc como un programa externo usando bun?

JJ reviewed (commented):
El principal problema es que no has hecho la revisión que te pedí; pero también creo que no has adaptado la orden a ejecutar al nuevo runtime.
Por cierto, ¿hace falta usar "bun run" para ejecutar cosas en un script de package.json?

vpedrosa commented on package.json:
```diff
@@ -0,0 +1,12 @@
+    "check": "bun run tsc --noEmit"
+  },
+  "devDependencies": {
+    "typescript": "^5.0.0"
```
Vale lo acabo de volver a mirar, typescript sí que lo necesitaría ya que el lenguaje elegido es typescript y para hacer el type check tengo que usarlo.

vpedrosa committed:
Refs #61, #62 - docs: elimino Turborepo y simplifico script check

- Elimino Turborepo del análisis de gestores de tareas (es un sistema
  de build para monorepos, no un gestor de tareas general)
- Simplifico script check eliminando 'bun run' redundante

vpedrosa committed:
Refs #61, #62 - docs: Modifico eliminando dependencias de typescript

	Cambio a tsc directamente desde bunx para no tener que utilizar
	typescript como devdependency.

vpedrosa committed:
Refs #61, #62 - ci: usa bun build nativo para check de sintaxis

Elimina dependencia de typescript usando bun build para verificar
sintaxis en lugar de tsc, ya que Bun transpila TypeScript nativamente.

vpedrosa commented:
Creo que con esto lo tengo listo @JJ, revísalo cuando puedas por favor.

JJ reviewed (approved):
👍

vpedrosa committed:
Refs #61, #62 - ci: simplifica check redirigiendo a /dev/null

Usa --outfile /dev/null en lugar de crear y eliminar un directorio
temporal, como sugirió el profesor.

vpedrosa commented on package.json:
```diff
@@ -0,0 +1,9 @@
+  "description": "Detector de buenas épocas de surf para playas",
+  "type": "module",
+  "scripts": {
+    "check": "bun build src/domain/index.ts --outdir /tmp/bun-check && rm -rf /tmp/bun-check"
```
Sí, efectivamente. Corrijo, y mezclo. Gracias!
